src/main/java/com/cleanroommc/modularui/widgets/RichTextWidget.java

```diff
@@ -55,83 +53,76 @@ public void drawForeground(ModularGuiContext context) {
+        if (getHoveredElement() instanceof Interactable interactable) {
             return interactable.onMouseTapped(mouseButton);
         }
         return Result.IGNORE;
     }
 
     @Override
     public @NotNull Result onKeyPressed(char typedChar, int keyCode) {
-        Object o = this.text.getHoveringElement(getContext().getFontRenderer(), getContext().getMouseX(), getContext().getMouseY());
-        if (o instanceof Interactable interactable) {
+        if (getHoveredElement() instanceof Interactable interactable) {
             return interactable.onKeyPressed(typedChar, keyCode);
         }
         return Result.ACCEPT;
     }
 
     @Override
     public boolean onKeyRelease(char typedChar, int keyCode) {
-        Object o = this.text.getHoveringElement(getContext().getFontRenderer(), getContext().getMouseX(), getContext().getMouseY());
-        if (o instanceof Interactable interactable) {
+        if (getHoveredElement() instanceof Interactable interactable) {
             return interactable.onKeyRelease(typedChar, keyCode);
         }
         return false;
     }
 
     @Override
     public @NotNull Result onKeyTapped(char typedChar, int keyCode) {
-        Object o = this.text.getHoveringElement(getContext().getFontRenderer(), getContext().getMouseX(), getContext().getMouseY());
-        if (o instanceof Interactable interactable) {
+        if (getHoveredElement() instanceof Interactable interactable) {
             return interactable.onKeyTapped(typedChar, keyCode);
         }
         return Result.ACCEPT;
     }
 
     @Override
     public boolean onMouseScroll(ModularScreen.UpOrDown scrollDirection, int amount) {
-        Object o = this.text.getHoveringElement(getContext().getFontRenderer(), getContext().getMouseX(), getContext().getMouseY());
-        if (o instanceof Interactable interactable) {
+        if (getHoveredElement() instanceof Interactable interactable) {
             return interactable.onMouseScroll(scrollDirection, amount);
         }
         return false;
     }
 
     @Override
     public void onMouseDrag(int mouseButton, long timeSinceClick) {
-        Object o = this.text.getHoveringElement(getContext().getFontRenderer(), getContext().getMouseX(), getContext().getMouseY());
-        if (o instanceof Interactable interactable) {
+        if (getHoveredElement() instanceof Interactable interactable) {
             interactable.onMouseDrag(mouseButton, timeSinceClick);
         }
     }
 
     @Nullable
     public Object getHoveredElement() {
+        return getHoveredElement(getContext());
+    }
+
+    @Nullable
+    public Object getHoveredElement(ModularGuiContext context) {
         if (!isHovering()) return null;
-        getContext().pushMatrix();
-        getContext().translate(getArea().x, getArea().y);
-        Object o = this.text.getHoveringElement(getContext());
-        getContext().popMatrix();
-        return o;
+        return this.text.getHoveringElement(context);
```

Comment:
why are you removing the transform

Reply:
because it's already translated for the interactable/drawing methods, translating again makes it stop working

Reply:
Im confused. It was working before, was it not? And you didnt change anything where it would translate.

Reply:
this method wasn't used previously

Reply:
nvm, this method was only used in ClientScreenHandler, but now i'm using it for the interactable/draw methods in the widget itself. using it as is caused problems with the matrix translate. ClientScreenHandler would need the translate, but the widget's internal methods don't